src/mlstacks/utils/terraform_utils.py

```diff
@@ -0,0 +1,621 @@
+            ret_code, _stdout, _stderr = client.apply(
+                var=tf_vars,
+                input=True,
+                capture_output=False,
+                raise_on_error=True,
+                refresh=False,
+                auto_approve=False,
+                skip_plan=False,
+            )
+        else:
+            ret_code, _stdout, _stderr = client.apply(
+                var=tf_vars,
+                input=False,
+                capture_output=True,
+                raise_on_error=False,
+                refresh=False,
+                skip_plan=True,
+                auto_approve=True,
+            )
+    except python_terraform.TerraformCommandError as e:
+        # TODO: pull all this error handling out to somewhere else
+        # TODO: catch the error!
+        if "The specified location constraint is not valid" in e:
+            logger.error(
+                f"The region '{tf_vars['region']}' you provided is invalid. "
+                "Please fix and try again."
+            )
+    logger.debug("Terraform changes successfully applied.")
+    return ret_code, _stdout, _stderr
+
+
+def tf_client_destroy(
+    client: python_terraform.Terraform, tf_vars: Dict[str, Any], debug: bool
+) -> Tuple[Any, Any, Any]:
+    """Destroy Terraform changes.
+
+    Args:
+        client: The Terraform client.
+        tf_vars: The Terraform variables.
+        debug: Whether to run in debug mode.
+
+    Returns:
+        The return code, stdout, and stderr.
+    """
+    logger.debug("Destroying Terraform components...")
+    if debug:
+        ret_code, _stdout, _stderr = client.destroy(
+            var=tf_vars,
+            input=True,
+            capture_output=False,
+            raise_on_error=True,
+            force=python_terraform.IsNotFlagged,
+            refresh=False,
+            auto_approve=False,
+            # skip_plan=False,
+        )
+    else:
+        ret_code, _stdout, _stderr = client.destroy(
+            var=tf_vars,
+            capture_output=True,
+            raise_on_error=False,
+            force=python_terraform.IsNotFlagged,
+            refresh=False,
+            auto_approve=True,
+            # skip_plan=True,
+        )
+    logger.debug("Terraform components successfully destroyed.")
+    return ret_code, _stdout, _stderr
+
+
+def clean_stack_recipes() -> None:
+    """Deletes stack recipe files from config directory."""
+    logger.info("Cleaning stack recipes...")
+    tf_path = f"{CONFIG_DIR}/terraform"
+    shutil.rmtree(tf_path)
+    logger.info(f"Deleted Terraform directory at {tf_path}")
+
+
+def deploy_stack(stack_path: str, debug_mode: bool = False) -> None:
+    """Deploy stack.
+
+    Args:
+        stack_path: The path to the stack.
+        debug_mode: Whether to run in debug mode.
+    """
+    # load and parse terraform variables and definitions
+    stack = load_stack_yaml(stack_path)
+    tf_recipe_path = f"{CONFIG_DIR}/terraform/{stack.provider}-modular"
```

Comment:
```suggestion
    tf_recipe_path = os.path.join(CONFIG_DIR, "terraform", f"{stack.provider}-modular")
```

Reply:
maybe you can also define this centrally somewhere so you don't need to construct these paths in every function

Reply:
or even refactor some of this into object-oriented might make sense